### PR TITLE
feat(lineage source): add fine grained lineage support

### PR DESCRIPTION
**Pull request by anshbansal**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/examples/mce_files/bootstrap_mce.json [hunk 1]
```diff
@@ -862,7 +862,7 @@
                 "upstreamType": "FIELD_SET",
                 "upstreams": ["urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:hive,logging_events,PROD),event_data)"],
                 "downstreamType": "FIELD",
-                "upstreams": ["urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:hive,fct_users_created,PROD),user_name)"],
+                "downstreams": ["urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:hive,fct_users_created,PROD),user_name)"],
```
Comment (hsheth2) [inline metadata-ingestion/examples/mce_files/bootstrap_mce.json:865]: yikes how long has this been like that

File: metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py [hunk 2]
```diff
@@ -49,9 +54,44 @@ def type_must_be_supported(cls, v: str) -> str:
...
+
+    @validator("upstreamType")
+    def upstream_type_must_be_supported(cls, v: str) -> str:
+        allowed_types = [
+            FineGrainedLineageUpstreamType.FIELD_SET,
+            FineGrainedLineageUpstreamType.DATASET,
+            FineGrainedLineageUpstreamType.NONE,
+        ]
+        if v not in allowed_types:
+            raise ConfigurationError(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py:73]: raise ValueError here

File: metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py [hunk 3]
```diff
@@ -49,9 +54,44 @@ def type_must_be_supported(cls, v: str) -> str:
...
+        return v
+
+    @validator("downstreamType")
+    def downstream_type_must_be_supported(cls, v: str) -> str:
+        allowed_types = [
+            FineGrainedLineageDownstreamType.FIELD_SET,
+            FineGrainedLineageDownstreamType.FIELD,
+        ]
+        if v not in allowed_types:
+            raise ConfigurationError(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py:85]:
```suggestion
            raise ValueError(
```

**Review by hsheth2 — COMMENTED** [review]
approach looks pretty reasonable

File: metadata-ingestion/examples/mce_files/bootstrap_mce.json (hunk 1 above)
Comment (anshbansal) [reply metadata-ingestion/examples/mce_files/bootstrap_mce.json:865]: Don't really know

File: metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py [hunk 4]
```diff
@@ -49,9 +53,44 @@ def type_must_be_supported(cls, v: str) -> str:
         return v
 
 
+class FineGrainedLineageConfig(ConfigModel):
+    upstreamType: str = "NONE"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py:57]: probably want a default of FIELD_SET

File: metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py [hunk 5]
```diff
@@ -49,9 +53,44 @@ def type_must_be_supported(cls, v: str) -> str:
...
+    upstreams: Optional[List[str]]
+    downstreamType: str
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py:59]: should default this to FIELD

File: metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py [hunk 6]
```diff
@@ -49,9 +53,44 @@ def type_must_be_supported(cls, v: str) -> str:
...
+    downstreams: Optional[List[str]]
+    transformOperation: Optional[str]
+    confidenceScore: Optional[float]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/lineage.py:62]: default to 1.0

**Review by hsheth2 — APPROVED** [review]
nits around the defaults but otherwise lgtm

**Review by hsheth2 — APPROVED** [review]

**Comment by anshbansal** [comment]
Waiting on https://github.com/datahub-project/datahub/pull/8135 for docs to get fixed

**Comment by anshbansal** [comment]
Doc build will be fixed in the linked PR.

**Comment by githendrik** [comment]
please excuse me hijacking this PR - I'm just wondering if there's a roadmap entry to allow querying the transformOperation via GraphQL and / or visualising it in the UI? We have a case for this specific feature where our users would like to see the actual operation in the UI to better understand the lineage.

Could someone maybe point me in the right direction if this is planned already? Otherwise happy to try and contribute back.

**Comment by hsheth2** [comment]
@githendrik we don't have any concrete plans for it, but I think it probably does make sense do to that.

Surfacing it via GraphQL feels like a no-brainer, and we'd definitely accept a PR there.

I'm less sure where we'd want to surface it in the UI - might be worth thinking about more / mocking. The reason is that it's an "edge annotation", but there's a bunch of other edge annotations that we might also want to support in the future e.g. the actual SQL logic, confidence scores on our lineage generation, etc

**Comment by hsheth2** [comment]
Surfacing it via graphql is a no-brainer - we'd definitely accept a PR for that

I'm less sure where to surface it in the UI - might be worth building some mocks for that. Specifically, there's a lot that we'd want to show as "edge annotations" - transform operation, the raw SQL logic, lineage confidence scores, where the lineage came from, created/updated timestamps per edge. I don't want to do something that we'd need to undo later.

**Pull request merged**